### PR TITLE
Rebased grabber

### PR DESCRIPTION
**Pull request by DmitryAstafyev**

### CODE REVIEW

File: application/apps/rustcore/rs-bindings/src/js/session.rs [hunk 1]
```diff
@@ -511,18 +540,23 @@ impl RustSession {
                 ));
             }
         };
-        self.op_channel
-            .0
-            .send(Operation::Assign {
+        let op_channel_tx = self.op_channel.0.clone();
+        async move {
+            match op_channel_tx.send(Operation::Assign {
```
Comment (marcmo) [inline application/apps/rustcore/rs-bindings/src/js/session.rs:545]: this does not look correct...what you are trying to do is to return a future. For that you should not artificially create an async block but rather use a combinator like `ready`
https://docs.rs/futures/0.3.15/futures/future/fn.ready.html

Comment (marcmo) [reply application/apps/rustcore/rs-bindings/src/js/session.rs:545]: Did you check what happens if you block in an async function? Is the js thread then also blocked? if so we need to wrap the blocking call in a `tokio::task::spawn_blocking`

File: application/apps/indexer/processor/src/search.rs [hunk 2]
```diff
@@ -208,6 +250,64 @@ impl SearchHolder {
...
+                .join("|")
+        );
+        let mut values: Vec<ExtractedMatchValue> = vec![];
+        let mut regexs: Vec<Regex> = vec![];
+        for filter in self.filters.iter() {
+            match Regex::from_str(&filter_as_regex(filter)) {
+                Ok(reg) => regexs.push(reg),
+                Err(err) => return Err(SearchError::Regex(format!("{}", err))),
+            }
+        }
```
Comment (marcmo) [inline application/apps/indexer/processor/src/search.rs:292]: much nicer: `Regex::from_string(...).map_err(|err| SearchError::Regex(...)?`

File: application/apps/indexer/processor/src/grabber.rs [hunk 3]
```diff
@@ -12,9 +12,12 @@ pub trait GrabTrait {
     fn grab_content(&self, line_range: &LineRange) -> Result<GrabbedContent, GrabError>;
     fn inject_metadata(&mut self, metadata: GrabMetadata) -> Result<(), GrabError>;
     fn get_metadata(&self) -> Option<&GrabMetadata>;
+    fn drop_metadata(&mut self);
     fn associated_file(&self) -> PathBuf;
 }
 
+pub trait AsyncGrabTrait: GrabTrait + Sync + Send + std::fmt::Debug {}
```
Comment (marcmo) [inline application/apps/indexer/processor/src/grabber.rs:19]: do we really need to add those traits with this commit? I thought you only need them for the cancellation stuff

File: application/apps/indexer/processor/src/search.rs [hunk 4]
```diff
@@ -8,8 +8,11 @@ use std::{
     fs::File,
     io::{BufWriter, Write},
     path::{Path, PathBuf},
+    str::FromStr,
 };
+use regex::Regex;
 use thiserror::Error;
+use indexer_base::{progress::ComputationResult, utils};
```
Comment (marcmo) [inline application/apps/indexer/processor/src/search.rs:15]: Pay attention to the rust linter warnings here and get rid of unused imports

File: application/apps/rustcore/rs-bindings/src/js/session.rs [hunk 5]
```diff
@@ -622,27 +657,68 @@ impl RustSession {
...
+        };
+        let filters: Vec<SearchFilter> = filters.iter().map(|f| f.as_filter()).collect();
+        info!(
+            "Extract (operation: {}) will be done in {:?} withing next filters: {:?}",
+            operation_id, target_file, filters
+        );
+        let op_channel_tx = self.op_channel.0.clone();
+        async move {
+            match op_channel_tx.send(Operation::Extract {
+                target_file,
```
Comment (marcmo) [inline application/apps/rustcore/rs-bindings/src/js/session.rs:706]: same issue as above

File: application/apps/rustcore/rs-bindings/src/js/session.rs (hunk 1 above)
Comment (DmitryAstafyev) [reply application/apps/rustcore/rs-bindings/src/js/session.rs:545]: you are absolutely right, that `async {}` block doesn't necessary here at all as soon as `#[node_bindgen]` macro cares about wrapping into the future.

File: application/apps/indexer/processor/src/grabber.rs (hunk 3 above)
Comment (DmitryAstafyev) [reply application/apps/indexer/processor/src/grabber.rs:19]: This trait related to async functionality, but not to cancelation.

**Pull request merged**